spotify_player/src/cli/client.rs

```diff
@@ -350,3 +365,425 @@ async fn handle_playback_request(
+                                    .await?
+                                    .as_str(),
+                            );
+                            result.push('\n');
+                        }
+                        Ok(result)
+                    } else {
+                        Err(anyhow!("No imports found for '{}'", to_id.id()))
+                    }
+                } else {
+                    Ok(format!("Not following '{}'", to_id.id()))
+                }
+            } else {
+                // Updating all imports
+
+                let mut result = String::new();
+
+                let dirs = imports_dir.read_dir()?;
+                for dir in dirs {
+                    let dir_path = dir?.path();
+                    let dir_name = dir_path.file_name().unwrap().to_str().unwrap();
+
+                    let to_id = PlaylistId::from_id(dir_name.to_owned())?;
+
+                    let pl_follow = client
+                        .spotify
+                        .playlist_check_follow(to_id.to_owned(), &[uid.to_owned()])
+                        .await?
+                        .pop()
+                        .unwrap();
+
+                    // No import for non following playlist
+                    if pl_follow {
+                        let to_dir = imports_dir.join(dir_name);
+                        for file in to_dir.read_dir()? {
+                            let file_name = file?.file_name().into_string().unwrap();
+
+                            let from_id = PlaylistId::from_id(file_name)?;
+
+                            result.push_str(
+                                playlist_import(client, from_id, to_id.to_owned())
+                                    .await?
+                                    .as_str(),
+                            );
+                            result.push('\n');
+                        }
+                    } else {
+                        // Remove non following imports as they are now useless
+                        remove_dir_all(dir_path)?;
+                        result.push_str(
+                            format!(
+                                "Not following playlist '{}'. Deleting import...\n",
+                                to_id.id()
+                            )
+                            .as_str(),
+                        )
+                    }
+                }
+
+                Ok(result)
+            }
+        }
+    }
+}
+
+/// Imports a playlist into another playlist.
+/// All tracks from the imported playlist are added to the import-to playlist if they are not in there already.
+/// The state of the imported playlist and a hash of the track ids are written to a file
+///
+/// This file is used in subsequent imports of the same two playlists
+/// It is first used to see if the state of the imported playlist has changed
+/// then if so, is used to see what has differed in the change.
+async fn playlist_import(
```

Comment:
Assuming that importing a `from` playlist to a `to` playlist means adding tracks that belong to `from` but don't belong to `to`, this function's implementation shouldn't be more than 20-30 lines. 

IMHO, this is kinda overcomplicated.

Reply:
I humbly disagree. It is slightly more than what it could be, but its slight increase in complexity reduces extra iterations and allows for important features. 

The logic follows 3 possible states:
- New import - add all tracks and write to file
- Old import, no change - check that hash is the same, then do nothing
- Old import, changes - add new tracks and write to file

Would there be a way for me to break this down to make it more simple and readable?

Reply:
> Assuming that importing a `from` playlist to a `to` playlist means adding tracks that belong to `from` but don't belong to `to`, this function's implementation shouldn't be more than 20-30 lines.

Another reason I forgot to mention is that I do not want to just add anything that is not in `to` but any changes in `from` into `to`. The difference is that I do not add anything that was already added before, meaning you can delete items that were added from an import and the subsequent imports will not add them back. This is a desired action.

For this to work, I cannot just request to see the state of `from`, but look to the file I wrote to see what has been added before.

I use this feature specifically. I am importing a friends playlist that I like but I don't like some of the songs, so I delete them. The deleted songs do not get re-added  but I get any new additions to my friends playlist.

Reply:
> Another reason I forgot to mention is that I do not want to just add anything that is not in `to` but any changes in `from` into `to`. The difference is that I do not add anything that was already added before, meaning you can delete items that were added from an import and the subsequent imports will not add them back. This is a desired action.
> 
> For this to work, I cannot just request to see the state of `from`, but look to the file I wrote to see what has been added before.
> 
> I use this feature specifically. I am importing a friends playlist that I like but I don't like some of the songs, so I delete them. The deleted songs do not get re-added but I get any new additions to my friends playlist.

Ah, now it makes more sense why you need to file logs. 

My point is that this adds another layer to the things you need to know when using the CLI commands. When looking at the `import`, I'll expect it to do one thing: importing all tracks from one playlist to another playlist without duplications.

From a user standpoint, when looking into the command description without looking into the implementation, I have no idea why there is `delete` option for `import` command and the meaning of the `update` command. That means you'll need to explain to users the concept of storing playlist import states into files and how we do it for users to be able to use those commands.

I mean it still comes down to a matter of personal preferences. I want the CLIs to be simple and just do one thing by default. You can build more complicated behavior by combining different CLI usages, hence writing scripts.

Reply:
Let me review this PR again to see if the implementation can be simplified and if it's possible to improve the documentation so that users don't need to know about the underlying import logic with the filesystem.

Reply:
> I humbly disagree. It is slightly more than what it could be, but its slight increase in complexity reduces extra iterations and allows for important features.
> 
> The logic follows 3 possible states:
> 
> * New import - add all tracks and write to file
> * Old import, no change - check that hash is the same, then do nothing
> * Old import, changes - add new tracks and write to file
> 
> Would there be a way for me to break this down to make it more simple and readable?

Ok, I understand the main idea of the `import` function now. Will give it a try to simplify the codes.

Reply:
> Ok, I understand the main idea of the import function now. Will give it a try to simplify the codes.

Created a PR in https://github.com/justjokiing/spotify-player/pull/1.